### PR TITLE
DOCS: Specifying Colors tutorial format & arrange

### PR DESCRIPTION
**Pull request by jeromefv**
## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: tutorials/colors/colors.py [hunk 1]
```diff
@@ -5,59 +5,85 @@
...
-How "Alpha" behaves depends on the ``zorder`` of the Artist.  Higher
-``zorder`` Artists are drawn on top of lower Artists, and "Alpha" determines
-whether the lower artist is covered by the higher.
-If the old RGB of a pixel is ``RGBold`` and the RGB of the
-pixel of the Artist being added is ``RGBnew`` with Alpha ``alpha``,
-then the RGB of the pixel is updated to:
-``RGB = RGBOld * (1 - Alpha) + RGBnew * Alpha``.  Alpha
-of 1 means the old color is completely covered by the new Artist, Alpha of 0
-means that pixel of the Artist is transparent.
+"Alpha" depends on the ``zorder`` of the Artist.  Matplotlib draws higher
```
Comment (story645) [inline tutorials/colors/colors.py:65]: reading through this whole section, I'm wondering if it'd be easier to explain via diagrams/mpl patches (and understand if this should be pitched to another issue/pr?) like two overlapping boxes - A1, A2 ... with A1 > A2 when z_a1>z_a2 and vice versa.  Then make a second example where alpha on A1 is very low, and possibly third example where A1 and A2 both have like .5 alpha.

**Review by story645 — CHANGES_REQUESTED** [review]
The table makes me very happy. I'm finding the RGB section a bit hard to follow but I don't know if my suggestions are in scope for this PR

File: tutorials/colors/colors.py (hunk 1 above)
Comment (jeromefv) [reply tutorials/colors/colors.py:65]: It's a good explanation and I think it's a valuable example to illustrate. Though I'm not entirely sure if it would make the most sense here in the colors tutorials section. Is there already a place with `zorder` information where this could be referenced? Or, is this something that could be a new section?

Comment (story645) [reply tutorials/colors/colors.py:65]: the zorder demo doesn't include color, https://matplotlib.org/gallery/misc/zorder_demo.html#sphx-glr-gallery-misc-zorder-demo-py
diagrams about color/zorder interaction: https://colab.research.google.com/drive/1U6I8yICGWvwwgOIgr7C_Hg7uoCDDD4OH#scrollTo=HesrKyf98bSX

Comment (tacaswell) [reply tutorials/colors/colors.py:65]: I would not say "alpha depends on zorder" as it is the two of them together that determine how the colors are blended.  Maybe something like:

> Alpha controls how the colors are combined when Artists overlap each other.  The Artists are rendered in order based on their zorder, Artists with higher zorder are "on top of" Artist with lower zorder, and combined via alpha compositing.    To compute the result of blending a new artist we use the equation ::
>   
>     RGB_{out} = RGB_{bottom} * (1 - \alpha) + RGB_{top} * \alpha
>
> If $\alpha$ is 1 the top color fully occludes the bottom color, if $\alpha$ is 0 the top color is not visible and they are blended at for intermediate values of $\alpha$.  Because each artist is blended into the cumulative result of all previous artists, re-ordering Artists is not commutative. 

xref: https://en.wikipedia.org/wiki/Alpha_compositing  ?

**Review by tacaswell — CHANGES_REQUESTED** [review]
I am good with all of the changes except for the alpha-related changes.

Anyone can clear this review after changes to the text related to alpha.

File: tutorials/colors/colors.py (hunk 1 above)
Comment (jklymak) [reply tutorials/colors/colors.py:65]: ^ I found that a *bit* confusing because "bottom" really means whatever is bellow the current artist, which itself may be the result of alpha compositing.

```
Artists are composed based on their zorder, followed by the order they are added to an axes.  The Alpha for an artist controls how the RGB color of the new artist is combined with the colors already on the axes using the equation ::
   RGB_{new} = RGB_{below} * (1-\alpha) + RGB_{artist} * \alpha
If $\alpha$ is 1 the top color fully occludes the bottom color, if $\alpha$ is 0 the top color is not visible, and they are blended for intermediate values.  
```
I find the last suggested sentence more mysterious than helpful in that I'm not sure what situation they could be confused for being commutative?

Comment (tacaswell) [reply tutorials/colors/colors.py:65]: I like @jklymak 's suggested names s better than mine.

I have a strong memory of cranking through the algebra to convince my self that it was non-commutative at one point so I am not sure it is completely obvious ;).   We have at least one bug report about this (https://github.com/matplotlib/matplotlib/issues/4862).

Comment (jklymak) [reply tutorials/colors/colors.py:65]: OK, I agree.  I wonder if this is the place to go into it in a bit more detail or not, but if so, it should be a paragraph with an example.

Comment (jeromefv) [reply tutorials/colors/colors.py:65]: I've changed the language of Alpha and Artists based on `zorder` for the section, though I'm not entirely sure how to include an example of the colors changing. Would placing the code sample from @story645 below the text be the best fit?

Comment (story645) [reply tutorials/colors/colors.py]: I dunno that you need the full code sample (which maybe stash in a file for reference & link to in docs). I'd just use the images to illustrate the idea (+/- a change to the title).

Comment (jeromefv) [reply tutorials/colors/colors.py]: Should I add a static file to the library and reference it in the paragraph? Also, does "Specifying Colors" need to change for the title or is the title change something else?

Comment (story645) [reply tutorials/colors/colors.py]: So yeah I think a static image would be fine...and my title change I meant the axes titles in the figure if they're not as clear as could be.

File: tutorials/colors/colors.py [hunk 2]
```diff
@@ -5,59 +5,104 @@
...
-How "Alpha" behaves depends on the ``zorder`` of the Artist.  Higher
-``zorder`` Artists are drawn on top of lower Artists, and "Alpha" determines
-whether the lower artist is covered by the higher.
-If the old RGB of a pixel is ``RGBold`` and the RGB of the
-pixel of the Artist being added is ``RGBnew`` with Alpha ``alpha``,
-then the RGB of the pixel is updated to:
-``RGB = RGBOld * (1 - Alpha) + RGBnew * Alpha``.  Alpha
-of 1 means the old color is completely covered by the new Artist, Alpha of 0
-means that pixel of the Artist is transparent.
+Matplotlib composes Artists based on ``zorder`` and then by the order added to
```
Comment (story645) [inline tutorials/colors/colors.py:65]: Maybe more explicit here?
Based on the "zorder" keyword argument and if not specified then by the order added

**Review by story645 — APPROVED** [review]

File: tutorials/colors/colors.py (hunk 2 above)
Comment (jeromefv) [reply tutorials/colors/colors.py]: No problem! I've modified the language to clarify the process.

File: tutorials/colors/colors.py [hunk 3]
```diff
@@ -5,59 +5,105 @@
...
+| RGB or RGBA (red, green, blue, alpha)| - ``(0.1, 0.2, 0.5)``                |
+| tuple of float values in a closed    | - ``(0.1, 0.2, 0.5, 0.3)``           |
+| interval [0, 1].                     |                                      |
++--------------------------------------+--------------------------------------+
+| Case-insensitive hex RGB or RGBA     | - ``'#0f0f0f'``                      |
+| string.                              | - ``'#0f0f0f80'``                    |
++--------------------------------------+--------------------------------------+
+| Case-insensitive shorthand equivalent| - ``'#abc'`` as ``'#aabbcc'``        |
+| string of RGB or RGBA from duplicated| - ``'#abcd'`` as ``'#aabbccdd'``     |
+| characters.                          |                                      |
```
Comment (QuLogic) [inline tutorials/colors/colors.py:20]: Reading this by itself, I'm not too sure what it means; I think a slight re-arrangement might work?
```
Case-insensitive shorthand RGB or RGBA string, made equivalent to above by duplicating characters.
```

File: tutorials/colors/colors.py [hunk 4]
```diff
@@ -5,59 +5,105 @@
...
++--------------------------------------+--------------------------------------+
+| String representation of float value | - ``'0.8'`` as light gray            |
+| in closed interval ``[0, 1]`` for    | - ``'0'`` as black                   |
+| black and white, respectively.       | - ``'1'`` as white                   |
++--------------------------------------+--------------------------------------+
+| Single character shorthand notation  | - ``'b'`` as blue                    |
+| for shades of colors.                | - ``'g'`` as green                   |
```
Comment (timhoffm) [inline tutorials/colors/colors.py:27]: Link to https://matplotlib.org/devdocs/gallery/color/named_colors.html

File: tutorials/colors/colors.py [hunk 5]
```diff
@@ -5,59 +5,105 @@
...
+|           magenta, and yellow do not | - ``'m'`` as magenta                 |
+|           coincide with X11/CSS4     | - ``'y'`` as yellow                  |
+|           colors.                    | - ``'k'`` as black                   |
+|                                      | - ``'w'`` as white                   |
++--------------------------------------+--------------------------------------+
+| Case-insensitive color name from     | - ``'xkcd:sky blue'``                |
+| `xkcd color survey`_ with ``'xkcd:'``| - ``'xkcd:eggshell'``                |
+| prefix.                              |                                      |
++--------------------------------------+--------------------------------------+
+| Case-insensitive Tableau Colors from | - 'tab:blue'                         |
```
Comment (timhoffm) [inline tutorials/colors/colors.py:39]: Link to https://matplotlib.org/devdocs/gallery/color/named_colors.html

File: tutorials/colors/colors.py [hunk 6]
```diff
@@ -5,59 +5,105 @@
...
+| Single character shorthand notation  | - ``'b'`` as blue                    |
+| for shades of colors.                | - ``'g'`` as green                   |
+|                                      | - ``'r'`` as red                     |
+| .. note:: The colors green, cyan,    | - ``'c'`` as cyan                    |
+|           magenta, and yellow do not | - ``'m'`` as magenta                 |
+|           coincide with X11/CSS4     | - ``'y'`` as yellow                  |
+|           colors.                    | - ``'k'`` as black                   |
+|                                      | - ``'w'`` as white                   |
++--------------------------------------+--------------------------------------+
+| Case-insensitive color name from     | - ``'xkcd:sky blue'``                |
```
Comment (timhoffm) [inline tutorials/colors/colors.py:35]: X11/CSS4 colors are missing.

and link to https://matplotlib.org/devdocs/gallery/color/named_colors.html

File: tutorials/colors/colors.py [hunk 7]
```diff
@@ -5,59 +5,105 @@
...
+|                                      | - 'tab:green'                        |
+|                                      | - 'tab:red'                          |
+|                                      | - 'tab:purple'                       |
+| .. note:: This is the default color  | - 'tab:brown'                        |
+|           cycle.                     | - 'tab:pink'                         |
+|                                      | - 'tab:gray'                         |
+|                                      | - 'tab:olive'                        |
+|                                      | - 'tab:cyan'                         |
++--------------------------------------+--------------------------------------+
+| "CN" color spec where ``'C'``        | - ``'C0'``                           |
```
Comment (timhoffm) [inline tutorials/colors/colors.py:50]: Link to "CN color selection" section below.

File: tutorials/colors/colors.py [hunk 8]
```diff
@@ -5,59 +5,105 @@
...
+| precedes a number acting as an index | - ``'C1'``                           |
+| into the default property cycle.     +--------------------------------------+
+|                                      | :rc:`axes.prop_cycle`                |
+| .. note:: Indexing occurs at         |                                      |
+|           rendering time and defaults|                                      |
```
Comment (timhoffm) [inline tutorials/colors/colors.py:55]: This seems to contradict

> "CN" colors convert to RGBA when creating Artists.

below.

File: tutorials/colors/colors.py [hunk 9]
```diff
@@ -5,59 +5,105 @@
...
-How "Alpha" behaves depends on the ``zorder`` of the Artist.  Higher
-``zorder`` Artists are drawn on top of lower Artists, and "Alpha" determines
-whether the lower artist is covered by the higher.
-If the old RGB of a pixel is ``RGBold`` and the RGB of the
-pixel of the Artist being added is ``RGBnew`` with Alpha ``alpha``,
-then the RGB of the pixel is updated to:
-``RGB = RGBOld * (1 - Alpha) + RGBnew * Alpha``.  Alpha
-of 1 means the old color is completely covered by the new Artist, Alpha of 0
-means that pixel of the Artist is transparent.
+Matplotlib composes Artists based on the ``zorder`` keyword argument. If there
```
Comment (timhoffm) [inline tutorials/colors/colors.py:65]: zorder only specifies the draw order (which one is on top). "Composing" is a bit too big a word here and a bit vague at the same time.

Also, I would use "parameter" instead of "keyword argument", because a keyword argument is only one way of setting the zorder; `set_zorder()` would be another way.

File: tutorials/colors/colors.py [hunk 10]
```diff
@@ -5,59 +5,105 @@
...
+are no specified values, Matplotlib defaults to the order of the Artists added
+to the Axes. The alpha for an Artist controls how the RGB color of the new
```
Comment (timhoffm) [inline tutorials/colors/colors.py:67]: Should mention that alpha specifies the transparency (or technically opaqueness?).

**Review by jeromefv — COMMENTED** [review]
I didn't mean to start my own review.

File: tutorials/colors/colors.py (hunk 3 above)
Comment (QuLogic) [reply tutorials/colors/colors.py]: 'shorthand' is really more of a modifier on 'string' or 'RGB or RGBA'; I don't like it in the latter half there. It reads strangely.

**Review by dstansby — APPROVED** [review]
This looks awesome 🥳

File: tutorials/colors/colors.py [hunk 11]
```diff
@@ -78,33 +131,34 @@ def demo(sty):
...
-# Out of 148 colors in the CSS color list, there are 95 name collisions
-# between the X11/CSS4 names and the xkcd names, all but 3 of which have
-# different hex values.  For example ``'blue'`` maps to ``'#0000FF'``
-# where as ``'xkcd:blue'`` maps to ``'#0343DF'``.  Due to these name
-# collisions all of the xkcd colors have ``'xkcd:'`` prefixed.  As noted in
-# the blog post, while it might be interesting to re-define the X11/CSS4 names
-# based on such a survey, we do not do so unilaterally.
+# There are 95 out of 148 colors with name collisions between the X11/CSS4
+# names and the xkcd names. Only three of these colors have the same hex
+# values.
```
Comment (timhoffm) [inline tutorials/colors/colors.py:154]: Maybe a bit less technical:

```suggestion
# 95 out of the 148 X11/CSS4 color names also appear in the xkcd survey.
# Almost all of them map to different color values in the X11/CSS4 and in
# the xkcd palette. Only 'black', 'white' and 'cyan' are identical.
```

**Review by timhoffm — APPROVED** [review]
This is a significant improvement! :+1: Let's get it in. One can always polish up more, but that can be done another time. :smile:
